### ISSUE TITLE
set_permissions wildcard expansion

### ISSUE THREAD

**Issue by marscher** [rabbitmq/rabbitmq-server#1331]
Hi,

I'd guess this one is trivial, but I can not figure out why protected (quoted) .* permissions always expands to file names.

I'm trying to invoke the set_permissions calls like in documented in the quick start guide:
```
rabbitmqctl set_permissions -p myvhost myuser ".*" ".*" ".*"
```

But the result is the following:
```
Error: operation set_permissions used with invalid parameter: ["myuser",
                                                               "'.*",".","..",
                                                               ".mkdir",".*'"]
```

Using version 3.6.10 installed by conda from conda-forge.

I've tried already lots of options of disable expansion in the invoking shell (bash), but somewhere this is "undone" again. I really wonder why this does not seem to be an issue to more people, because expansion is a default feature of this commonly used shell. Please enlighten me! Thanks a lot.

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
Your shell expands it. Why it may be expanding quoted values, I'm not sure.

**Comment by lukebakken** [comment]
@marscher - I think there's a hint in the error output:

```
"myuser",
"'.*",
".",
"..",
".mkdir",
".*'"
```

Notice that the second element in that list is `"'.*"` - with a single quote, and that the last element is `".*'"`. I recommend double-checking that you are using actual double-quote characters and haven't copied something like U+201C and U+201D.

Or, change your quoting to use single-quotes:

```
rabbitmqctl set_permissions -p myvhost myuser '.*' '.*' '.*'
```

**Comment by marscher** [comment]
Thank you for your quick responses. I've double checked, that there are no extra quotes, just using single quotes also lead to file name expansion.

**Comment by marscher** [comment]
I ended up using the web interface to set the permissions...

**Comment by lukebakken** [comment]
OK. As @michaelklishin said, there must be something awry with your shell or terminal. If this were an issue with `rabbitmqctl` be assured we would hear about it non-stop as `.*` is a very common regex.

**Comment by marscher** [comment]
On 08/11/2017 05:59 PM, Luke Bakken wrote:
> OK. As @michaelklishin <https://github.com/michaelklishin> said, there
> must be something awry with your shell or terminal. If this were an
> issue with |rabbitmqctl| be assured we would hear about it non-stop as
> |.*| is a very common regex.

yes, it is very likely that my shell is borked.

**Comment by costrouc** [comment]
Have the exact same issue with a the conda forge package

**Comment by pcm32** [comment]
I also have the issue with the conda forge based installation... I get a weird expansion error:

```bash
$ rabbitmqctl set_permissions -p vhost user ".*" ".*" ".*"
Error (argument validation): too many arguments.
Arguments given:
	set_permissions -p vhost user . .. .mkdir . .. .mkdir . .. .mkdir

Usage

rabbitmqctl [--node <node>] [--longnames] [--quiet] set_permissions [--vhost <vhost>] <username> <conf> <write> <read>
```

**Comment by pcm32** [comment]
However if I do:
```
echo ".*"
.*
```
so I don't think that my shell is expanding the asterisk...

**Comment by pcm32** [comment]
```
[{pid,13761},
 {running_applications,
     [{rabbit,"RabbitMQ","3.7.16"},
      {os_mon,"CPO  CXC 138 46","2.4.7"},
      {sysmon_handler,"Rate-limiting system_monitor event handler","1.1.0"},
      {rabbit_common,
          "Modules shared by rabbitmq-server and rabbitmq-erlang-client",
          "3.7.16"},
      {ranch,"Socket acceptor pool for TCP protocols.","1.7.1"},
      {ssl,"Erlang/OTP SSL application","9.2"},
      {public_key,"Public key infrastructure","1.6.5"},
      {asn1,"The Erlang ASN1 compiler version 5.0.8","5.0.8"},
      {crypto,"CRYPTO","4.4.1"},
      {observer_cli,"Visualize Erlang Nodes On The Command Line","1.5.0"},
      {jsx,"a streaming, evented json parsing toolkit","2.9.0"},
      {mnesia,"MNESIA  CXC 138 12","4.15.6"},
      {recon,"Diagnostic tools for production use","2.5.0"},
      {inets,"INETS  CXC 138 49","7.0.6"},
      {xmerl,"XML parser","1.3.19"},
      {lager,"Erlang logging framework","3.6.10"},
      {goldrush,"Erlang event stream processor","0.1.9"},
      {compiler,"ERTS  CXC 138 10","7.3.2"},
      {syntax_tools,"Syntax tools","2.1.7"},
      {sasl,"SASL  CXC 138 11","3.3"},
      {stdlib,"ERTS  CXC 138 10","3.8"},
      {kernel,"ERTS  CXC 138 10","6.3"}]},
 {os,{unix,linux}},
 {erlang_version,
     "Erlang/OTP 21 [erts-10.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:64] [hipe]\n"},
 {memory,
     [{connection_readers,0},
      {connection_writers,0},
      {connection_channels,0},
      {connection_other,0},
      {queue_procs,0},
      {queue_slave_procs,0},
      {plugins,17948},
      {other_proc,20904680},
      {metrics,195572},
      {mgmt_db,0},
      {mnesia,78096},
      {other_ets,3006136},
      {binary,2305168},
      {msg_index,65632},
      {code,24666620},
      {atom,1090753},
      {other_system,11473211},
      {allocated_unused,11558488},
      {reserved_unallocated,0},
      {strategy,rss},
      {total,[{erlang,63803816},{rss,74563584},{allocated,75362304}]}]},
 {alarms,[]},
 {listeners,[{clustering,25672,"::"},{amqp,5672,"::"}]},
 {vm_memory_calculation_strategy,rss},
 {vm_memory_high_watermark,0.4},
 {vm_memory_limit,3274371891},
 {disk_free_limit,50000000},
 {disk_free,293877237284864},
 {file_descriptors,
     [{total_limit,924},{total_used,4},{sockets_limit,829},{sockets_used,0}]},
 {processes,[{limit,1048576},{used,235}]},
 {run_queue,1},
 {uptime,110950},
 {kernel,{net_ticktime,60}}]
```

**Comment by pcm32** [comment]
@michaelklishin could we re-open this issue? Thanks!

**Comment by lukebakken** [comment]
@pcm32 I suspect that the "conda forge package" installs a script wrapper for `rabbitmqctl`. Can you grab that file and _**attach**_ it to your next comment here?

**Comment by pcm32** [comment]
Here is the wrapper attached
[conda_rabbitmqctl_wrapper.sh.txt](https://github.com/rabbitmq/rabbitmq-server/files/4712616/conda_rabbitmqctl_wrapper.sh.txt), thanks @lukebakken for taking a look.

**Comment by pcm32** [comment]
Setting a `set -o noglob` on line 28 right before the last exec sorts it out, for this at least, not sure if it would have consequences for other commands where you might want to use the `*`.

**Comment by lukebakken** [comment]
Glad that worked for you. That is a bug in the "conda package". If you have time, you should let them know.

**Comment by sysadmin-imran** [comment]
Below commands worked for me
```
kubectl exec -it $(kubectl get pods -o=name | grep rabbitmq ) -- rabbitmqctl -- add_user testuser testpassword

kubectl exec -it $(kubectl get pods -o=name | grep rabbitmq ) -- rabbitmqctl  set_permissions -p  my_vhost testuser "" ".*" ".*"

```

**Comment by Anthchirp** [comment]
> Glad that worked for you. That is a bug in the "conda package". If you have time, you should let them know.

Yes, I just had to re-discover this issue myself :smile: 
This should now be fixed in conda-forge rabbitmq 3.8.9-2 via https://github.com/conda-forge/rabbitmq-server-feedstock/pull/31

**Comment by costrouc** [comment]
@Anthchirp thanks for the PR. This was a critical issue for a client I was working with. Can verify that 3.8.9-2 fixes the issue.
